scripts/bo_benchmarks/demonstrations/branin_campaign_demonstration.py

```diff
@@ -0,0 +1,359 @@
+        mean_pred, std_pred = gp_model.predict(train_X)
+        
+        gp_r2 = r2_score(train_Y.detach().cpu().numpy(), mean_pred.detach().cpu().numpy())
+        
+        ls, imp = gp_model.get_lengthscales()
+        imp_dict = importance_concentration(imp)
+        
+        loo = loo_pseudo_likelihood(gp_model.model, train_X, train_Y)
+        
+        rank_tau = kendalltau(train_Y.squeeze().detach().cpu().numpy(), mean_pred.detach().cpu().numpy())[0]
+        
+        # Calculate interval score using GP predictions
+        y_true = train_Y.squeeze().detach().cpu().numpy()
+        y_pred_mean = mean_pred.detach().cpu().numpy()
+        y_pred_std = std_pred.detach().cpu().numpy()
+        
+        # Convert to 95% confidence intervals
+        lower, upper = normal_95ci(y_pred_mean, y_pred_std)
+        interval_scores = interval_score(y_true, lower, upper)
+        mean_interval_score = np.mean(interval_scores)
+        
+        # ===== Ax cross validation metrics =====
+        ax_cv_r2, ax_cv_interval_score = calculate_ax_cross_validation_metrics(ax_client)
+        
+        metrics = {
+            'trial': trial_index,
+            'gp_r2': gp_r2,  # GP-based R²
+            'ax_cv_r2': ax_cv_r2,  # Ax cross validation R²
+            'interval_score': mean_interval_score,  # Mean interval score (GP-based)
+            'ax_cv_interval_score': ax_cv_interval_score,  # Mean interval score (Ax CV)
+            'imp_cumsum': imp_dict,
+            'loo_nll': loo,
+            'rank_tau': rank_tau,
+            'lengthscales': ls,
+            'importance': imp
+        }
+        
+        return metrics
+        
+    except Exception as e:
+        print(f"Metrics calculation failed: {e}")
+        return None
+
+
+# Create ax client
+ax_client = AxClient()
+
+ax_client.create_experiment(
+    parameters=[
+        {"name": "x1", "type": "range", "bounds": [-5.0, 10.0]},
+        {"name": "x2", "type": "range", "bounds": [0.0, 10.0]},
+    ],
+    objectives={
+        obj1_name: ObjectiveProperties(minimize=True),
+    },
+)
+
+# Track metrics per iteration
+all_metrics = []
+
+# Run 10 trials with iteration-by-iteration metrics calculation
+print("Running Branin optimization campaign with 10 trials...")
+for i in range(10):
+    parameterization, trial_index = ax_client.get_next_trial()
+
+    # extract parameters
+    x1 = parameterization["x1"]
+    x2 = parameterization["x2"]
+
+    results = branin(x1, x2)
+    ax_client.complete_trial(trial_index=trial_index, raw_data=results)
+    
+    print(f"Trial {trial_index}: x1={x1:.3f}, x2={x2:.3f}, branin={results:.3f}")
+    
+    # Calculate metrics after each trial (from trial 3 onward)
+    iteration_metrics = calculate_iteration_metrics(ax_client, trial_index)
+    if iteration_metrics:
+        all_metrics.append(iteration_metrics)
+        print(f"  GP R²: {iteration_metrics['gp_r2']:.4f}, LOO NLL: {iteration_metrics['loo_nll']:.2f}")
```

Comment:
Maybe all these print statements would be better served in the metrics calculation functions? To preserve clarity and separation in this optimization campaign loop